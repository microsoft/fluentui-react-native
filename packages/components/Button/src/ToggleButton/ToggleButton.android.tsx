/** @jsx withSlots */
import { View } from 'react-native';
import { toggleButtonName, ToggleButtonType } from './ToggleButton.types';
import { TextV1 as Text } from '@fluentui-react-native/text';
import { compose } from '@fluentui-react-native/framework';
import { Icon } from '@fluentui-react-native/icon';

export const ToggleButton = compose<ToggleButtonType>({
  displayName: toggleButtonName,
  slots: {
    root: View,
    icon: Icon,
    content: Text,
  },
  useRender: () => {
    return () => {
      console.warn('Toggle Button is not implemented for Android');
      return null;
    };
  },
});
