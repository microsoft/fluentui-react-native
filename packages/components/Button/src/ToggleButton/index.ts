export * from './ToggleButton.types';
export * from './ToggleButton';
