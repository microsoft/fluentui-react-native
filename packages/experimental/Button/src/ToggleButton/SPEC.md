# Toggle Button

## Background

The `ToggleButton` is a `Button` that is designed show a selected state when toggled.

## Sample Code

Basic examples:

```jsx
<ToggleButton>Text</ToggleButton>
<ToggleButton checked={true}>Text</ToggleButton>
<ToggleButton defaultChecked={true}>Text</ToggleButton>
<ToggleButton icon={{ svgSource: { uri: 'https://www.example.com/test.svg', viewBox: '0 0 100 100' } }} />
<ToggleButton icon={{ svgSource: { uri: 'https://www.example.com/test.svg', viewBox: '0 0 100 100' } }}>Text</ToggleButton>
<ToggleButton appearance="primary">Text</ToggleButton>
<ToggleButton disabled>Text</ToggleButton>
<ToggleButton size="small">Text</ToggleButton>
```

## Visual Examples

Win32:

![ToggleButton with text on win32 example](../../assets/togglebutton_example_win32.png)

```tsx
<ToggleButton>Text</ToggleButton>
```

![ToggleButton with text and checked state on win32 example](../../assets/togglebutton_checked_example_win32.png)

```tsx
<ToggleButton checked={true}>Text</ToggleButton>
```

## Variants

Variant options are the same as the base `Button` component.

## API

### Slots

- `root` - The outer container representing the `Button` itself that wraps everything passed via the `children` prop.
- `icon` - If specified, renders an `icon` either before or after the `children` as specified by the `iconPosition` prop.
- `loader` - If specified, renders a `loader` before `children` while the `loading` flag is set to `true` in place of an icon. This slot is mutually exclusive to the icon slot.

### Props

```ts
export interface ToggleButtonProps extends ButtonProps {
  /**
   * Defines the controlled checked state of the `ToggleButton`.
   * Mutually exclusive to `defaultChecked`.
   * This should only be used if the checked state is to be controlled at a higher level and there is a plan to pass the
   * correct value based on handling `onClick` events and re-rendering.
   */
  checked?: boolean;

  /**
   * Defines whether the `ToggleButton` is inititally in a checked state or not when rendered.
   * Mutually exclusive to `checked`.
   */
  defaultChecked?: boolean;
}
```

### Styling Tokens

```ts
export interface CompoundButtonTokens extends ButtonTokens {
  /**
   * States that can be applied to a button.
   * These can be used to modify styles of the button when under the specified state.
   */
  checked?: ToggleButtonTokens;
}
```

## Behaviors

This button includes the same behaviors as base `Button`. It has some additional states.

### States

The following section describes the additional states a `ToggleButton` can have.

#### Checked state

A checked `ToggleButton` changes styling to communicate that the button is currently selected or toggled.
