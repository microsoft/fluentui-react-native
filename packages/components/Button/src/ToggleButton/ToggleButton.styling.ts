import type { Theme, UseStylingOptions } from '@fluentui-react-native/framework';
import { buildProps } from '@fluentui-react-native/framework';
import { borderStyles, layoutStyles, fontStyles } from '@fluentui-react-native/tokens';

import { toggleButtonName } from './ToggleButton.types';
import type { ToggleButtonTokens, ToggleButtonSlotProps, ToggleButtonProps } from './ToggleButton.types';
import { buttonStates, contentStyling } from '../Button.styling';

export const stylingSettings: UseStylingOptions<ToggleButtonProps, ToggleButtonSlotProps, ToggleButtonTokens> = {
  tokens: [toggleButtonName],
  states: ['checked', ...buttonStates],
  slotProps: {
    root: buildProps(
      (tokens: ToggleButtonTokens, theme: Theme) => ({
        style: {
          display: 'flex',
          alignItems: 'center',
          flexDirection: 'row',
          alignSelf: 'flex-start',
          justifyContent: 'center',
          width: tokens.width,
          backgroundColor: tokens.backgroundColor,
          ...borderStyles.from(tokens, theme),
          ...layoutStyles.from(tokens, theme),
        },
      }),
      ['backgroundColor', 'width', ...borderStyles.keys, ...layoutStyles.keys],
    ),
    content: buildProps(
      (tokens: ToggleButtonTokens, theme: Theme) => ({
        style: {
          ...contentStyling(tokens, theme, tokens.color, tokens),
        },
      }),
      ['color', 'spacingIconContentAfter', 'spacingIconContentBefore', ...fontStyles.keys],
    ),
    icon: buildProps(
      (tokens: ToggleButtonTokens) => ({
        style: {
          tintColor: tokens.iconColor,
        },
        height: tokens.iconSize,
        width: tokens.iconSize,
      }),
      ['iconColor', 'iconSize'],
    ),
    focusInnerBorder: buildProps(
      (tokens: ToggleButtonTokens) => ({
        style: {
          position: 'absolute',
          borderWidth: tokens.borderInnerWidth,
          borderColor: tokens.borderInnerColor,
          borderRadius: tokens.borderInnerRadius,
        },
      }),
      ['borderInnerWidth', 'borderInnerColor', 'borderInnerRadius'],
    ),
  },
};
