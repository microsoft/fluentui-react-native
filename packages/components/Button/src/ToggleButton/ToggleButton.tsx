/** @jsx withSlots */
import * as React from 'react';
import { Pressable } from 'react-native';
import { ActivityIndicator } from '@fluentui-react-native/experimental-activity-indicator';
import { ToggleButtonProps, toggleButtonName, ToggleButtonType } from './ToggleButton.types';
import { TextV1 as Text } from '@fluentui-react-native/text';
import { stylingSettings } from './ToggleButton.styling';
import { compose, mergeProps, withSlots, UseSlots } from '@fluentui-react-native/framework';
import { Icon } from '@fluentui-react-native/icon';
import { createIconProps } from '@fluentui-react-native/interactive-hooks';
import { buttonLookup } from '../Button';
import { useToggleButton } from './useToggleButton';

export const ToggleButton = compose<ToggleButtonType>({
  displayName: toggleButtonName,
  ...stylingSettings,
  slots: {
    root: Pressable,
    icon: Icon,
    content: Text,
  },
  useRender: (userProps: ToggleButtonProps, useSlots: UseSlots<ToggleButtonType>) => {
    const iconProps = createIconProps(userProps.icon);
    const toggleButton = useToggleButton(userProps);

    // grab the styled slots
    const Slots = useSlots(userProps, (layer) => buttonLookup(layer, toggleButton.state, userProps));

    // now return the handler for finishing render
    return (final: ToggleButtonProps, ...children: React.ReactNode[]) => {
      const { icon, iconPosition, iconOnly, loading, accessibilityLabel, ...mergedProps } = mergeProps(toggleButton.props, final);
      const shouldShowIcon = !loading && icon;

      if (__DEV__ && iconOnly) {
        React.Children.forEach(children, (child) => {
          if (typeof child === 'string') {
            console.warn('iconOnly should not be set when Button has content.');
          }
        });
      }

      let childText = '';
      if (accessibilityLabel === undefined) {
        React.Children.forEach(children, (child) => {
          if (typeof child === 'string') {
            childText = child; // We only automatically support the one child string.
          }
        });
      }
      const label = accessibilityLabel ?? childText;

      return (
        <Slots.root {...mergedProps} accessibilityLabel={label}>
          {loading && <ActivityIndicator />}
          {shouldShowIcon && iconPosition === 'before' && <Slots.icon {...iconProps} />}
          {React.Children.map(children, (child) =>
            typeof child === 'string' ? <Slots.content key="content">{child}</Slots.content> : child,
          )}
          {shouldShowIcon && iconPosition === 'after' && <Slots.icon {...iconProps} />}
        </Slots.root>
      );
    };
  },
});
