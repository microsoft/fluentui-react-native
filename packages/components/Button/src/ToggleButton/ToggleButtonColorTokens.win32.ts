import { Theme } from '@fluentui-react-native/framework';
import { TokenSettings } from '@fluentui-react-native/use-styling';
import { ToggleButtonTokens } from './ToggleButton.types';

export const defaultToggleButtonColorTokens: TokenSettings<ToggleButtonTokens, Theme> = (t: Theme): ToggleButtonTokens => ({
  checked: {
    color: t.colors.neutralForeground1Selected,
    backgroundColor: t.colors.neutralBackground1Selected,
    borderColor: t.colors.neutralStroke1,
    primary: {
      color: t.colors.neutralForegroundOnBrandSelected,
      backgroundColor: t.colors.brandBackgroundSelected,
      borderColor: t.colors.brandBackgroundSelected,
    },
    subtle: {
      color: t.colors.neutralForeground1Selected,
      backgroundColor: t.colors.subtleBackgroundSelected,
      borderColor: t.colors.subtleBackgroundSelected,
    },
  },
});
