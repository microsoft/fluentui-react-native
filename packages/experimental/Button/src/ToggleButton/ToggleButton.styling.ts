import { toggleButtonName, ToggleButtonTokens, ToggleButtonSlotProps, ToggleButtonProps } from './ToggleButton.types';
import { Theme, UseStylingOptions, buildProps } from '@fluentui-react-native/framework';
import { borderStyles, fontStyles } from '@fluentui-react-native/tokens';
import { buttonStates, defaultButtonTokens } from '../ButtonTokens';

export const stylingSettings: UseStylingOptions<ToggleButtonProps, ToggleButtonSlotProps, ToggleButtonTokens> = {
  tokens: [
    defaultButtonTokens,
    (t: Theme) =>
      ({
        checked: {
          color: t.colors.buttonCheckedContent,
          backgroundColor: t.colors.buttonCheckedBackground,
          hovered: {
            color: t.colors.buttonCheckedHoveredContent,
            backgroundColor: t.colors.buttonCheckedHoveredBackground,
          },
          ghost: {
            color: t.colors.ghostCheckedContent,
            backgroundColor: t.colors.ghostCheckedBackground,
            hovered: {
              color: t.colors.ghostCheckedHoveredContent,
              backgroundColor: t.colors.ghostCheckedHoveredBackground,
              borderColor: t.colors.ghostCheckedHoveredBorder,
            },
          },
        },
      } as ToggleButtonTokens),
    toggleButtonName,
  ],
  states: ['checked', ...buttonStates],
  slotProps: {
    root: buildProps(
      (tokens: ToggleButtonTokens, theme: Theme) => ({
        style: {
          display: 'flex',
          alignItems: 'center',
          flexDirection: 'row',
          alignSelf: 'flex-start',
          justifyContent: 'center',
          minHeight: 32,
          minWidth: 80,
          width: tokens.width,
          paddingStart: 16,
          paddingEnd: 16,
          backgroundColor: tokens.backgroundColor,
          ...borderStyles.from(tokens, theme),
        },
      }),
      ['backgroundColor', ...borderStyles.keys],
    ),
    content: buildProps(
      (tokens: ToggleButtonTokens, theme: Theme) => ({
        style: {
          color: tokens.color,
          ...fontStyles.from(tokens, theme),
        },
      }),
      ['color', ...fontStyles.keys],
    ),
    icon: buildProps(
      (tokens: ToggleButtonTokens) => ({
        style: {
          tintColor: tokens.iconColor,
        },
      }),
      ['iconColor'],
    ),
  },
};
