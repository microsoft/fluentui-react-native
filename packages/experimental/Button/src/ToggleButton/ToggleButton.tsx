/** @jsx withSlots */
import * as React from 'react';
import { Image, View } from 'react-native';
import { ToggleButtonProps, toggleButtonName, ToggleButtonType } from './ToggleButton.types';
import { Text } from '@fluentui-react-native/experimental-text';
import { stylingSettings } from './ToggleButton.styling';
import { filterImageProps } from '@fluentui-react-native/adapters';
import { compose, mergeProps, withSlots, UseSlots } from '@fluentui-react-native/framework';
import { useButton } from '../useButton';
import { useAsToggle } from '@fluentui-react-native/interactive-hooks';

export const ToggleButton = compose<ToggleButtonType>({
  displayName: toggleButtonName,
  ...stylingSettings,
  slots: {
    root: View,
    icon: Image,
    content: Text,
  },
  filters: {
    icon: filterImageProps,
  },
  render: (userProps: ToggleButtonProps, useSlots: UseSlots<ToggleButtonType>) => {
    const { icon, content, defaultChecked, checked, onClick, ...rest } = userProps;

    // Warns defaultChecked and checked being mutually exclusive.
    if (defaultChecked != undefined && checked != undefined) {
      console.warn('defaultChecked and checked are mutually exclusive to one another. Use one or the other.');
    }
    const [checkedValue, toggle] = useAsToggle(defaultChecked, checked, onClick);
    const button = useButton({ onClick: toggle, ...rest });

    // grab the styled slots
    const Slots = useSlots(userProps, (layer) => (layer === 'checked' && checkedValue) || button.state[layer] || userProps[layer]);

    // now return the handler for finishing render
    return (final: ToggleButtonProps, ...children: React.ReactNode[]) => {
      const mergedProps = mergeProps(button.props, final);

      return (
        <Slots.root {...mergedProps}>
          {icon && <Slots.icon key="icon" source={{ uri: icon }} />}
          {content && <Slots.content key="content">{content}</Slots.content>}
          {children}
        </Slots.root>
      );
    };
  },
});

export default ToggleButton;
