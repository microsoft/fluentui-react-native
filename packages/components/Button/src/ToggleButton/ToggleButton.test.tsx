import * as React from 'react';

import { checkRenderConsistency, checkReRender } from '@fluentui-react-native/test-tools';
import * as renderer from 'react-test-renderer';

import { ToggleButton } from './ToggleButton';

it('ToggleButton default', () => {
  const tree = renderer.create(<ToggleButton>Default Button</ToggleButton>).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Button simple rendering does not invalidate styling', () => {
  checkRenderConsistency(() => <ToggleButton>Default button</ToggleButton>, 2);
});

it('Button re-renders correctly', () => {
  checkReRender(() => <ToggleButton>Render twice</ToggleButton>, 2);
});

it('Button shares produced styles across multiple renders', () => {
  const style = { backgroundColor: 'black' };
  checkRenderConsistency(() => <ToggleButton style={style}>Shared styles</ToggleButton>, 2);
});

it('Button re-renders correctly with style', () => {
  const style = { borderColor: 'blue' };
  checkReRender(() => <ToggleButton style={style}>Shared Style Render</ToggleButton>, 2);
});
