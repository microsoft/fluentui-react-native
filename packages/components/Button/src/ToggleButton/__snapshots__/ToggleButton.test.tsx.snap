// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleButton default 1`] = `
<View
  accessibilityActions={
    [
      {
        "name": "Toggle",
      },
    ]
  }
  accessibilityLabel="Default Button"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  enableFocusRing={true}
  focusable={true}
  onAccessibilityAction={[Function]}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#0f6cbd",
      "borderColor": "#0f6cbd",
      "borderRadius": 8,
      "borderWidth": 1,
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 40,
      "minWidth": 96,
      "paddingHorizontal": 12,
      "width": undefined,
    }
  }
>
  <Text
    accessible={false}
    ellipsizeMode="tail"
    numberOfLines={0}
    style={
      {
        "color": "#ffffff",
        "fontFamily": "System",
        "fontSize": 12,
        "fontWeight": "600",
        "margin": 0,
        "marginBottom": 0,
        "marginEnd": 0,
        "marginStart": 0,
        "marginTop": 0,
      }
    }
  >
    Default Button
  </Text>
</View>
`;
